text/3536-unsized-thin-pointers.md

```diff
@@ -0,0 +1,286 @@
+
+This pattern is used frequently in zero-copy APIs that transmit structured data
+between address spaces of differing trust levels.
+
+# Background
+[motivation]: #motivation
+
+There are currently two approved RFCs that cover similar functionality:
+* [RFC 1861] adds `extern type` for declaring types that are opaque to Rust's
+  type system. One of the capabilities available to extern types is that they
+  can be embedded into a `struct` as the last field, and that `struct` will
+  become an unsized type with thin references.
+
+  Stabilizing `extern type` is currently blocked on questions of how to handle
+  Rust layout intrinsics such as [`core::mem::size_of_val()`][size_of_val] and
+  [`core::mem::align_of_val()`][align_of_val] for fully opaque types.
+
+* [RFC 2580] adds traits and intrinsics for custom DSTs either with or without
+  associated "fat pointer" metadata. A custom DST with thin references can be
+  represented as `Pointee<Metadata = ()>`.
+
+  Stabilizing custom DSTs is currently blocked on multiple questions involving
+  the content and representation of complex metadata, such as `&dyn` vtables.
+
+In both of these cases the ability to declare custom DSTs with thin references
+is a minor footnote to the overall feature, and stabilization is blocked by
+issues unrelated to thin-pointer DSTs.
+
+The objective of this RFC is to extract custom thin-pointer DSTs into its own
+feature, which would hopefully be free of known issues and could be stabilized
+without significant changes to the compiler or ecosystem.
+
+[RFC 1861]: https://rust-lang.github.io/rfcs/1861-extern-types.html
+[RFC 2580]: https://rust-lang.github.io/rfcs/2580-ptr-meta.html
+
+[align_of_val]: https://doc.rust-lang.org/core/mem/fn.align_of_val.html
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The unsafe trait `core::mem::DynSized` may be implemented for a `!Sized` type
+to configure how the size of a value is computed from a reference. References
+to a type that implements `DynSized` are not required to store the value size
+as pointer metadata.
+
+If a type that implements `DynSized` has no other associated pointer metadata
+(such as a vtable), then references to that type will have the same size and
+layout as a normal pointer.
+
+```rust
+#[repr(C, align(8))]
+struct Request {
+	size: u32,
+	id: u16,
+	flags: u16,
+	data: [u8],
+}
+
+unsafe impl core::mem::DynSized for Request {
+	fn size_of_val(&self) -> usize {
+		usize::try_from(self.size).unwrap_or(usize::MAX)
+	}
+}
+
+// size_of::<&Request>() == size_of::<*const ()>()
+```
+
+The `DynSized` trait has a single required method, `size_of_val()`, which
+has the same semantics as `core::mem::size_of_val()`.
+
+```rust
+// core::mem
+pub unsafe trait DynSized {
+	// Returns the size of the pointed-to value in bytes.
+	fn size_of_val(&self) -> usize;
+}
+```
+
+It is an error to `impl DynSized` for a type that is `Sized`. In other words,
+the following code is invalid:
```

Comment:
I don't want to allow code like that. The behavior of `size_of_val()` on an `UnknownVariant` would be ambiguous.

Restricting `DynSized` to be `!Sized` means that there's no question about which definition of "value size" is being requested -- it's the one determinable from the reference.

Reply:
I would expect code something like this:

```rust
#[repr(C)]
struct Header {
    variant_id: u32,
}

#[repr(C)]
struct Request {
    header: Header,
    data: [u8],
}
unsafe impl DynSized for Request { ... }


impl Request {
    fn header(&self) -> &Header { &self.header }
    fn variant(&self) -> Variant<'_> { ... }
}

enum Variant<'a> {
    A(&'a VariantA),
    B(&'b VariantB),
    Unknown,
}

#[repr(C)]
struct VariantA {
  header: Header,
  foo: u8,
}

#[repr(C)]
struct VariantB {
  header: Header,
  bar: f64,
}
```

The `Request` struct is `!Sized`, each of the variant structs are `Sized` -- in this example you basically have a tagged `union` where you've reserved the right to have future variants contain `[u8]`.

Reply:
If you wanted to support something like this, maybe something like:
```
struct Thing {
    // header bits
   variant_id: u8,

   u: Variants
}

union Variants {
    f: f32,
    i: i32,
    t: AnotherThing
}
```
and treat it a bit like an enum with user defined layout, where the fat pointer metadata would encode which union field is active (with a "non selected" case for when you've yet to determine the variant). Implicit in that is that the dynamic size is header + size of selected variant field, rather than size of the overall union.

(Obviously today that would be treated as a normal struct with a union field, so you'd need some way to indicate this magic DST variant thing.)

Reply:
I might be misunderstanding, but I don't see how that would work. Consider the case of an IPv4 packet -- the data isn't any particular format, it's just `[u8]`, and the packet header can be used to discover the data length.

Your solution would imply a structure like this:

```rust
#[repr(C)]
struct Packet {
  header: PacketHeader,
  u: Variants,
}

union Variants {
  len0: [u8; 0],
  len1: [u8; 1],
  len2: [u8; 2],
  // ...
}
```

Reply:
Sorry that was a digression that doesn't have much to do with this rfc.